### DISCUSSION TITLE
Is it possible to host theia-front end and back-end on two different machines?

### DISCUSSION THREAD

**Discussion by kumbham** [eclipse-theia/theia#8827] [Q&A] [answered]
I read that theia has separate processes for backend and frontend. I'm curious if we could host these processes in two different machines and open a WebSocket connection for JSONRpc communication? If yes, could you please share a sample application with this configuration? 

It seems like these processes can also communicate via Rest APIs. Could you please share a code snippet for this configuration as well. 

Thanks!

**Comment by kittaakos** [comment, marked as answer]
> I read that theia has separate processes for backend and frontend.

Yes. The frontend process is the web-app associated with the browser's tab. The backend process can live on a separate machine that you access when you load the frontend application.

> I'm curious if we could host these processes in two different machines and open a WebSocket connection for JSONRpc communication?

Sure. You can try Theia in Gitpod: https://gitpod.io/#https://github.com/eclipse-theia/theia
The web app runs in your browser, and the Theia backend is running one of the nodes provided by Gitpod. They're on two different machines.

> It seems like these processes can also communicate via Rest APIs. Could you please share a code snippet for this configuration as well.

Yes, you can do this as well. For example, the file download functionality is exposed as a RESTful API using `express`. See an example [here](https://github.com/eclipse-theia/theia/blob/1892af580c0e28b5348919ddcd19979de082d20d/packages/filesystem/src/node/download/file-download-endpoint.ts#L44-L61).
